modules/sqs/main.tf

```diff
@@ -53,17 +53,81 @@ resource "aws_sqs_queue" "metadata_queue" {
   ## OPTIONAL
-  name                        = "${var.prefix}-staged-recovery-queue.fifo"
-  fifo_queue                  = true
-  content_based_deduplication = true
-  delay_seconds               = var.sqs_delay_time_seconds
-  max_message_size            = var.sqs_maximum_message_size
-  message_retention_seconds   = var.staged_recovery_queue_message_retention_time_seconds
-  tags                        = local.tags
-  policy                      = data.aws_iam_policy_document.staged_recovery_queue_policy.json
-  visibility_timeout_seconds  = 1800 # Set to double lambda max time
+  name                       = "${var.prefix}-staged-recovery-queue"
+  delay_seconds              = var.sqs_delay_time_seconds
+  max_message_size           = var.sqs_maximum_message_size
+  message_retention_seconds  = var.staged_recovery_queue_message_retention_time_seconds
+  tags                       = local.tags
+  policy                     = data.aws_iam_policy_document.staged_recovery_queue_policy.json
+  visibility_timeout_seconds = 1800 # Set to double lambda max time
+  depends_on = [
+    aws_sqs_queue.staged_recovery_deadletter_queue
+  ]
+  redrive_policy = jsonencode({
+    deadLetterTargetArn = aws_sqs_queue.staged_recovery_deadletter_queue.arn,
+    maxReceiveCount     = 3 #if number of delayed messages in the queue is greater than this number, then the messages will be sent to its DLQ.
+  })
 }
 
+resource "aws_sqs_queue" "staged_recovery_deadletter_queue" {
+  name                       = "${var.prefix}-staged-recovery-deadletter-queue"
+  delay_seconds              = var.sqs_delay_time_seconds
+  max_message_size           = var.sqs_maximum_message_size
+  message_retention_seconds  = var.staged_recovery_queue_message_retention_time_seconds
+  tags                       = local.tags
+  visibility_timeout_seconds = 1800 # Set to same time as the staged recovery queue.
+}
+
+resource "aws_sqs_queue_policy" "deadletter_queue_policy" {
+  queue_url = aws_sqs_queue.staged_recovery_deadletter_queue.id
+  policy    = data.aws_iam_policy_document.staged_recovery_deadletter_queue_policy.json
+}
+
+data "aws_iam_policy_document" "staged_recovery_deadletter_queue_policy" {
+  statement {
+    effect    = "Allow"
+    resources = [aws_sqs_queue.staged_recovery_deadletter_queue.arn]
+    actions = [
+      "sqs:ChangeMessageVisibility",
+      "sqs:DeleteMessage",
+      "sqs:GetQueueAttributes",
+      "sqs:GetQueueUrl",
+      "sqs:ListQueueTags",
+      "sqs:ReceiveMessage",
+      "sqs:SendMessage",
+    ]
+  }
+}
+
+resource "aws_cloudwatch_metric_alarm" "staged_recovery_deadletter_alarm" {
+  alarm_name          = "${aws_sqs_queue.staged_recovery_deadletter_queue.name}-not-empty-alarm"
+  alarm_description   = "Items are on the ${aws_sqs_queue.staged_recovery_deadletter_queue.name} queue"
+  comparison_operator = "GreaterThanOrEqualToThreshold"
+  evaluation_periods  = 1
+  metric_name         = "ApproximateNumberOfMessagesVisible"
```

Comment:
Might vote for NumberOfMessagesReceived, with a comparison_operator of sum. That way messages stop being received if the queue stops filling up. As-is, emails will be sent out once per period for as long as the queue remains non-empty.

Reply:
yes `NumberOfMessagesReceived` should also work. That could be an alternative.

From my testing, everything looked good with the above config. I only get one alarm if there are message in DLQ.
In your suggested config, I am assuming we will continue getting alarm emails every time period which might fill our inbox.

Reply:
Is there something built into the alarm that stops emails from coming in? As written:
If the queue has more than one element, trigger alarm.
What I'm suggesting:
If the queue receives more than one element, trigger alarm.
The former will trigger for as long as the queue remains non-empty, the later will trigger only when new items are added to the queue. Unless there's something where this alarm can only trigger emails once until the queue is empty again.

Reply:
right now, if the number of messages in DLQ is `GreaterThanOrEqualToThreshold ` which is 1, it will send an alarm only once.
when the queue is empty again and then it receives a message, it will again send an alarm

Reply:
Ah, gotcha. Thanks for the explanation.